Implement design system buttons

steff456 opened the pull request:
Part of https://github.com/conda-incubator/conda-store-ui/issues/288
<!-- Reference the issue corresponding to this PR. If an issues does not exist, consider opening one or writing a detailed description for what this PR changes and it's value in the following sections. -->
<!-- GitHub Docs on Keywords: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests -->

## Description
<!-- What is the purpose of this pull request? -->

This pull request:

- Implements buttons following the design system

### Main buttons

**Default**

<img width="94" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/188abfc4-564e-4715-8f24-685088cf5e33">

**Hover**

<img width="79" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/efd17879-8f37-45f7-8bb0-1a6badc6b97c">

**Focus**

<img width="101" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/11beb9f8-e99f-47ac-9296-35720da27c62">

**Disabled**

<img width="147" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/cc5fa7af-7559-4c30-b10c-bfa2f0267652">

### Icon buttons

**Default**

<img width="48" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/81f1009c-8034-4b6a-bbd0-7c82e9574dc3">

**Hover**

<img width="48" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/931890d3-1e49-423f-a143-f7d332dc8639">

**Focus**

<img width="49" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/d8199750-e4a3-4ee3-bdd4-5277ef30c844">

### Toggle switch

**Default**

<img width="166" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/7eec55f5-b9b8-46ee-9936-b5b91eee3ca5">

**Hover** / **Focus**

<img width="166" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/d1c45eec-1dbc-4c5c-9106-433e2e9fba86">

## Pull request checklist
<!-- Quick checklist to ensure high-quality Pull Request. -->

- [x] Did you test this change locally?
- [x] Did you update the documentation (if required)?
- [x] Did you add/update relevant tests for this change (if required)?

## Additional information
<!-- Do you have any other information about this pull request? This may include screenshots, references, and/or implementation notes. -->

- [x] This PR depends on first merging https://github.com/conda-incubator/conda-store-ui/pull/292

### Screenshots
 
<details>
#### Sidebar

**Before**
<img width="294" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/c6f123a6-8ddc-446a-ac0a-6c2b79ee9fa5">

**After**
<img width="292" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/1443c6f3-40ba-49e8-902e-c8ee026d5ae6">

#### View Environment

**Before**
<img width="562" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/f341a475-05c2-4eca-8603-8e3a3c90357e">

**After**
<img width="521" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/ab1e1e41-95b7-4b49-8710-a3d9c06e5263">

#### Edit Environment

**Before**
<img width="566" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/8d68c8ba-1b68-4e35-a666-3ca403c6068f">

<img width="594" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/2ac333c3-1a07-4f85-b87c-86dfed3b3cc1">

**After**
<img width="516" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/f5cb6aa4-7fb5-4773-b593-1e21647f1b5c">

<img width="519" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/dd92d6b7-8b77-4b83-b228-a892742dde5a">

#### Create Environment

**Before**
<img width="603" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/0bf5aa33-9d90-45a0-9de3-bc99dbc8cd33">

**After**
<img width="958" alt="image" src="https://github.com/conda-incubator/conda-store-ui/assets/20992645/257e705e-258c-43b7-b73e-4dcd934770dc">

trallard reviewed (changes requested):
Quick notes:

* View environment: there is a slight shade difference (now) between the first environment and the rest. Is this a hover state? If so, this might need tweaking as the change is barely noticeable

* Edit environment: 
	* The `Requested packages` section has a green border - it should be gray
	* The toggle button (white) almost blends with the background. There is a shadow, but it is not enough to make it stand out. Can we add a border instead? @smeragoel
	* Same with the active env dropdown, this has an extra green border

* There is not enough contrast on the disabled buttons (text vs background) - thought disabled buttons would be gray? @smeragoel 

![Colour_Contrast_Analyser__CCA__and_Implement_design_system_buttons_by_steff456_·_Pull_Request__304_·_conda-incubator_conda-store-ui](https://github.com/conda-incubator/conda-store-ui/assets/23552331/eb1ea086-bc48-444b-b725-70c0375bf494)


* I am +1 on underlining links and items on focus - though the underline is correct against the text. Can we add an underlined offset?

This is what we have on PST, and it works well because it scales accordingly.

```scss
// Define some useful variables for links styling consistency
//
//  Thickness of the underline for links
// the default will be either:
//  - 1px
// - 0.0625rem if it's thicker than 1px because the user has changed the text
//   size in their browser
$link-underline-thickness: unquote("max(1px, .0625rem)") !default;
// Offset of link underlines from text baseline
// The default is 3px expressed as ems, as calculated against the default body
// font size (on desktop).
$link-underline-offset: 0.1578em !default;
```

This is based on the screenshots @steff456, as I only had a glance at the code. But from this quick glance it looks already much better in terms of maintainability ✨  thank you

smeragoel commented:
> * View environment: there is a slight shade difference (now) between the first environment and the rest. Is this a hover state? If so, this might need tweaking as the change is barely noticeable

I didn't add this component to the design system, but I will.

>   * The toggle button (white) almost blends with the background. There is a shadow, but it is not enough to make it stand out. Can we add a border instead? @smeragoel

Okay, I will tweak the design so it's visible.

> * There is not enough contrast on the disabled buttons (text vs background) - thought disabled buttons would be gray? @smeragoel

They were supposed to be, but because of the greyscale design version, I used a lighter green. Let me work on a fix.
> 
> ![Colour_Contrast_Analyser__CCA__and_Implement_design_system_buttons_by_steff456_·Pull_Request__304·_conda-incubator_conda-store-ui](https://user-images.githubusercontent.com/23552331/270463001-eb1ea086-bc48-444b-b725-70c0375bf494.png)
> 
> * I am +1 on underlining links and items on focus - though the underline is correct against the text. Can we add an underlined offset?

I have added the offset to the designs as well!

trallard commented:
I think we can keep the disabled buttons gray here since this is the "default theme" and deal with the gray theme as the customisation it is

smeragoel commented:
Here's the new design for the toggle switch: https://www.figma.com/file/nKifk755TFMsK2SF082igS/conda-store-UI-design?type=design&node-id=1390%3A1571&mode=design&t=Pp6IQLDsLqaq4wps-1

Right now the toggle switch implies off/on states, which is not the most accurate signal to send, since we are just switching between interaction modes. I looked at examples from https://mui.com/material-ui/react-switch/ and used them as inspiration for the switch.

I also added clearer labels and interaction states. The icons (especially the gear for UI) and the content of the labels (GUI?) is up for discussion.

![image](https://github.com/conda-incubator/conda-store-ui/assets/98317216/2215ea11-5c9a-498b-a5eb-3edd00901dac)

trallard commented:
> I also added clearer labels and interaction states. The icons (especially the gear for UI) and the content of the labels (GUI?) is up for discussion.

I like these a lot - I prefer the one with the `< >` as the `gear` sends my mind straight to settings-like which might not be entirely accurate 🤷🏽‍♀️ 

`GUI` and `YAML` seem perfectly fine as labels

steff456 commented:
The latest commit fixes the disabled state and the hover style for buttons using the new guidelines. 

It has some work in progress for the switch button, but this PR still needs more changes that I will work on after coming back from PTO.

steff456 commented:
@trallard @smeragoel I just pushed the changes that address all the comments. I also updated the screenshots of all the views in the description and added the switch toggle states to it. 

I just want to note that I'm unsure on the style of the dropdown menu style, so this PR only changed the style of the `down arrow` button that is part of the component. We can iterate on the style of the actual list in a follow up PR.

This PR is ready for another review!

kcpevey commented:
I think this should be reviewed by @smeragoel before merging.

steff456 commented:
> One thing I noticed in the sidebar is that the bullets and the environment names are in green when it should be dark grey (secondary-800) by default.

This PR is only focusing on the buttons, I'm currently working in a follow up PR for styling the sidebar as you described in the comment.
